UOP1_Project/Assets/Scripts/Pool/Pool.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using OP1.Factory;
+
+namespace OP1.Pool
+{
+	/// <summary>
+	/// A generic pool that generates members of type T on-demand via a factory.
+	/// </summary>
+	/// <typeparam name="T">Specifies the type of elements in the pool.</typeparam>
+	public class Pool<T> : IEnumerable where T : IPoolable
+	{
+		public Stack<T> _available = new Stack<T>();
+		public HashSet<T> _unavailable = new HashSet<T>();
```

Comment:
So the reasoning here is that by maintaining a reference to unavailable objects, objects created outside the pool could not be added to the pool via Return.

However, I'm seeing now that I forgot to do that `_unavailable.Contains` check I meant to do in Return, so that's why that the collection looks useless.

Reply:
I see! Although now I'm thinking, is it necessary to have that constraint? We could have a system in which object `A` sends `Foo` to object `B`, `B` pools `Foo` to then be able to send it to `C` or back to `A` or whatever? We would be capable to reuse the same instance when possible. And if there's a case in which `B` needs to send out one before `A`, it'd be capable of spawning it on its own. 
Thinking as I type, that'd be solved by having a shared pool for `A` and `B` :P

But my main question came from, what if `Foo` gets destroyed or lost reference of before returning it? GC would never run on it because of the reference inside the pool, yet it'd never be reused. I guess that'd be a misusage of the pool but it'd cause leaks that would be hard to track.

Reply:
As much as I liked the idea of a closed system that could only pool the objects it made, you're right. The possibility of leaks outweighs the benefit of maintaining the reference. I've made the change. Pool now only has an available Stack.